test/extended/networking/livemigration.go

```diff
@@ -1,14 +1,680 @@
+								role:               "primary",
+								allowPersistentIPs: true,
+							},
+							fedoraVMIWithPrimaryUDNAttachment,
+							migrateVM,
+						),
+						Entry(
+							"when the VM attached to a primary UDN is restarted",
+							networkAttachmentConfigParams{
+								name:               nadName,
+								topology:           "layer2",
+								role:               "primary",
+								allowPersistentIPs: true,
+							},
+							fedoraVMWithPrimaryUDNAttachment,
+							restartVM,
+						),
+						Entry(
+							"when the VM attached to a secondary UDN is migrated between nodes",
+							networkAttachmentConfigParams{
+								name:               nadName,
+								topology:           "layer2",
+								role:               "secondary",
+								allowPersistentIPs: true,
+							},
+							fedoraVMWithSecondaryNetworkAttachment,
+							migrateVM,
+						),
+						Entry(
+							"when the VMI attached to a secondary UDN is migrated between nodes",
+							networkAttachmentConfigParams{
+								name:               nadName,
+								topology:           "layer2",
+								role:               "secondary",
+								allowPersistentIPs: true,
+							},
+							fedoraVMIWithSecondaryNetworkAttachment,
+							migrateVM,
+						),
+						Entry(
+							"when the VM attached to a secondary UDN is restarted",
+							networkAttachmentConfigParams{
+								name:               nadName,
+								topology:           "layer2",
+								role:               "secondary",
+								allowPersistentIPs: true,
+							},
+							fedoraVMWithSecondaryNetworkAttachment,
+							restartVM,
+						))
+				},
+				Entry("NetworkAttachmentDefinitions", func(c networkAttachmentConfigParams) {
+					netConfig := newNetworkAttachmentConfig(c)
+					nad := generateNAD(netConfig)
+					_, err := nadClient.NetworkAttachmentDefinitions(c.namespace).Create(context.Background(), nad, metav1.CreateOptions{})
+					Expect(err).To(Not(HaveOccurred()))
+				}),
+				Entry("UserDefinedNetwork", func(c networkAttachmentConfigParams) {
+					udnManifest := generateUserDefinedNetworkManifest(&c)
+					Expect(applyManifest(c.namespace, udnManifest)).To(Succeed())
+					Expect(waitForUserDefinedNetworkReady(c.namespace, c.name, udnCrReadyTimeout)).To(Succeed())
+				}))
 		})
 	})
 })
+
+type vmClient struct {
+	oc      *exutil.CLI
+	virtCtl string
+}
+
+func newVirtClient(oc *exutil.CLI, tmpDir string) *vmClient {
+	GinkgoHelper()
+	virtCtl, err := ensureVirtctl(oc, tmpDir)
+	Expect(err).To(Not(HaveOccurred()))
+	return &vmClient{
+		oc:      oc,
+		virtCtl: virtCtl,
+	}
+}
+
+func (v *vmClient) apply(description, resource string) error {
+	GinkgoHelper()
+	By(fmt.Sprintf("Create %s", description))
+	return applyManifest(v.oc.Namespace(), resource)
+}
+
+func (v *vmClient) virtctl(args []string) (string, error) {
+	GinkgoHelper()
+	output, err := exec.Command(v.virtCtl, args...).CombinedOutput()
+	if err != nil {
+		return "", err
+	}
+	return string(output), nil
+}
```

Comment:
I guess we could make this use the pattern of the rest of the tests and just use `Expect(err).NotTo(HaveOccurred())`

Reply:
I prefer not "polute" those helper with Expect so it's easier to use them at eventually and such.